File: reports/20260416-164621/results/storage/data-plane/java/crud/storage-dp-java-crud/baseline/claude-opus-4.6/generated-code/src/main/java/com/example/BlobStorageCrud.java
package com.example;

import com.azure.identity.DefaultAzureCredentialBuilder;
import com.azure.storage.blob.BlobClient;
import com.azure.storage.blob.BlobContainerClient;
import com.azure.storage.blob.BlobServiceClient;
import com.azure.storage.blob.BlobServiceClientBuilder;
import com.azure.storage.blob.models.BlobItem;
import com.azure.storage.blob.models.BlobStorageException;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

/**
 * Demonstrates CRUD operations on Azure Blob Storage using DefaultAzureCredential.
 *
 * <p>Set the environment variable {@code AZURE_STORAGE_ACCOUNT_NAME} to your storage account name
 * before running this application.</p>
 */
public class BlobStorageCrud {

    private static final String CONTAINER_NAME = "my-container";
    private static final String BLOB_NAME = "uploads/data.txt";
    private static final String LOCAL_FILE_PATH = "data.txt";
    private static final String DOWNLOADED_FILE_PATH = "data-downloaded.txt";

    public static void main(String[] args) {
        String accountName = System.getenv("AZURE_STORAGE_ACCOUNT_NAME");
        if (accountName == null || accountName.isBlank()) {
            System.err.println("Error: set the AZURE_STORAGE_ACCOUNT_NAME environment variable.");
            System.exit(1);
        }

        String endpoint = String.format("https://%s.blob.core.windows.net", accountName);

        // 1. Create a BlobServiceClient using DefaultAzureCredential
        BlobServiceClient serviceClient = new BlobServiceClientBuilder()
                .endpoint(endpoint)
                .credential(new DefaultAzureCredentialBuilder().build())
                .buildClient();
        System.out.println("Connected to storage account: " + accountName);

        try {
            createContainer(serviceClient);
            uploadBlob(serviceClient);
            listBlobs(serviceClient);
            downloadBlob(serviceClient);
            deleteBlob(serviceClient);
            deleteContainer(serviceClient);

            System.out.println("\nAll CRUD operations completed successfully.");
        } catch (BlobStorageException e) {
            System.err.printf("Azure Blob Storage error (HTTP %d): %s%n",
                    e.getStatusCode(), e.getServiceMessage());
        } catch (UncheckedIOException | IOException e) {
            System.err.println("I/O error: " + e.getMessage());
        }
    }

    /** 2. Create a container named "my-container" if it doesn't exist. */
    private static void createContainer(BlobServiceClient serviceClient) {
        System.out.println("\n--- Creating container ---");
        BlobContainerClient containerClient = serviceClient.getBlobContainerClient(CONTAINER_NAME);
        try {
            containerClient.create();
            System.out.println("Container created: " + CONTAINER_NAME);
        } catch (BlobStorageException e) {
            if (e.getStatusCode() == 409) {
                System.out.println("Container already exists: " + CONTAINER_NAME);
            } else {
                throw e;
            }
        }
    }

    /** 3. Upload a local file "data.txt" as a blob named "uploads/data.txt". */
    private static void uploadBlob(BlobServiceClient serviceClient) throws IOException {
        System.out.println("\n--- Uploading blob ---");

        Path localPath = Paths.get(LOCAL_FILE_PATH);
        if (!Files.exists(localPath)) {
            System.out.println("Local file not found; creating sample " + LOCAL_FILE_PATH);
            Files.writeString(localPath, "Hello, Azure Blob Storage!\n");
        }

        BlobClient blobClient = serviceClient
                .getBlobContainerClient(CONTAINER_NAME)
                .getBlobClient(BLOB_NAME);

        blobClient.uploadFromFile(localPath.toString(), true);
        System.out.printf("Uploaded '%s' as blob '%s'%n", LOCAL_FILE_PATH, BLOB_NAME);
    }

    /** 4. List all blobs in the container and print their names and sizes. */
    private static void listBlobs(BlobServiceClient serviceClient) {
        System.out.println("\n--- Listing blobs ---");
        BlobContainerClient containerClient = serviceClient.getBlobContainerClient(CONTAINER_NAME);

        for (BlobItem blobItem : containerClient.listBlobs()) {
            long size = blobItem.getProperties().getContentLength();
            System.out.printf("  Name: %-30s Size: %d bytes%n", blobItem.getName(), size);
        }
    }

    /** 5. Download the blob back to a local file "data-downloaded.txt". */
    private static void downloadBlob(BlobServiceClient serviceClient) {
        System.out.println("\n--- Downloading blob ---");
        BlobClient blobClient = serviceClient
                .getBlobContainerClient(CONTAINER_NAME)
                .getBlobClient(BLOB_NAME);

        blobClient.downloadToFile(DOWNLOADED_FILE_PATH, true);
        System.out.printf("Downloaded blob '%s' to '%s'%n", BLOB_NAME, DOWNLOADED_FILE_PATH);
    }

    /** 6. Delete the blob and then delete the container. */
    private static void deleteBlob(BlobServiceClient serviceClient) {
        System.out.println("\n--- Deleting blob ---");
        BlobClient blobClient = serviceClient
                .getBlobContainerClient(CONTAINER_NAME)
                .getBlobClient(BLOB_NAME);

        try {
            blobClient.delete();
            System.out.println("Deleted blob: " + BLOB_NAME);
        } catch (BlobStorageException e) {
            if (e.getStatusCode() == 404) {
                System.out.println("Blob not found (already deleted): " + BLOB_NAME);
            } else {
                throw e;
            }
        }
    }

    private static void deleteContainer(BlobServiceClient serviceClient) {
        System.out.println("\n--- Deleting container ---");
        BlobContainerClient containerClient = serviceClient.getBlobContainerClient(CONTAINER_NAME);

        try {
            containerClient.delete();
            System.out.println("Deleted container: " + CONTAINER_NAME);
        } catch (BlobStorageException e) {
            if (e.getStatusCode() == 404) {
                System.out.println("Container not found (already deleted): " + CONTAINER_NAME);
            } else {
                throw e;
            }
        }
    }
}

File: reports/20260501-181917/results/storage/data-plane/js-ts/crud/storage-dp-js-ts-crud/azure-mcp-skills/claude-sonnet-4.5/generated-code/blob-storage-crud.ts
import { BlobServiceClient, RestError } from "@azure/storage-blob";
import { DefaultAzureCredential } from "@azure/identity";

/**
 * Helper function to convert a readable stream to text
 */
async function streamToText(readable: NodeJS.ReadableStream): Promise<string> {
  const chunks: Buffer[] = [];
  for await (const chunk of readable) {
    chunks.push(Buffer.from(chunk));
  }
  return Buffer.concat(chunks).toString("utf-8");
}

/**
 * Main function demonstrating CRUD operations on Azure Blob Storage
 */
async function main() {
  try {
    // 1. Create a BlobServiceClient using DefaultAzureCredential
    console.log("1. Creating BlobServiceClient...");
    const accountName = process.env.AZURE_STORAGE_ACCOUNT_NAME;
    
    if (!accountName) {
      throw new Error("AZURE_STORAGE_ACCOUNT_NAME environment variable is required");
    }

    const credential = new DefaultAzureCredential();
    const blobServiceClient = new BlobServiceClient(
      `https://${accountName}.blob.core.windows.net`,
      credential
    );
    console.log("✓ BlobServiceClient created successfully\n");

    // 2. Create a container named "my-container" if it doesn't exist
    console.log("2. Creating container 'my-container'...");
    const containerClient = blobServiceClient.getContainerClient("my-container");
    
    const createContainerResponse = await containerClient.createIfNotExists();
    if (createContainerResponse.succeeded) {
      console.log("✓ Container created successfully");
    } else {
      console.log("✓ Container already exists");
    }
    console.log();

    // 3. Upload a string "Hello Azure!" as a block blob named "greeting.txt"
    console.log("3. Uploading blob 'greeting.txt'...");
    const blobName = "greeting.txt";
    const content = "Hello Azure!";
    const blockBlobClient = containerClient.getBlockBlobClient(blobName);
    
    await blockBlobClient.upload(content, content.length, {
      blobHTTPHeaders: { blobContentType: "text/plain" }
    });
    console.log(`✓ Uploaded '${blobName}' successfully\n`);

    // 4. List all blobs in the container and log their names
    console.log("4. Listing all blobs in the container:");
    let blobCount = 0;
    for await (const blob of containerClient.listBlobsFlat()) {
      console.log(`  - ${blob.name} (${blob.properties.contentLength} bytes)`);
      blobCount++;
    }
    console.log(`✓ Found ${blobCount} blob(s)\n`);

    // 5. Download the blob and print its content as a string
    console.log("5. Downloading blob 'greeting.txt'...");
    const blobClient = containerClient.getBlobClient(blobName);
    const downloadResponse = await blobClient.download();
    
    if (!downloadResponse.readableStreamBody) {
      throw new Error("Failed to download blob: no readable stream");
    }

    const downloadedContent = await streamToText(downloadResponse.readableStreamBody);
    console.log(`✓ Downloaded content: "${downloadedContent}"\n`);

    // 6. Delete the blob and then delete the container
    console.log("6. Cleaning up resources...");
    
    // Delete the blob
    await blobClient.delete();
    console.log(`✓ Deleted blob '${blobName}'`);
    
    // Delete the container
    await containerClient.delete();
    console.log("✓ Deleted container 'my-container'\n");

    console.log("✅ All CRUD operations completed successfully!");

  } catch (error) {
    // Error handling with RestError
    if (error instanceof RestError) {
      console.error("\n❌ Azure Storage error occurred:");
      console.error(`   Status Code: ${error.statusCode}`);
      console.error(`   Error Code: ${error.code}`);
      console.error(`   Message: ${error.message}`);
      
      // Provide specific guidance based on error code
      switch (error.statusCode) {
        case 403:
          console.error("\n💡 Tip: Check that your credentials have the required permissions:");
          console.error("   - Storage Blob Data Contributor role or higher");
          break;
        case 404:
          console.error("\n💡 Tip: The resource was not found. Check the account name and container/blob names.");
          break;
        case 409:
          console.error("\n💡 Tip: Resource already exists or is being deleted.");
          break;
        default:
          console.error("\n💡 Tip: Check your Azure Storage account configuration and network connectivity.");
      }
    } else if (error instanceof Error) {
      console.error("\n❌ Error:", error.message);
    } else {
      console.error("\n❌ Unknown error:", error);
    }
    
    process.exit(1);
  }
}

// Run the main function
main();
